distance-method/distance_method.py

```diff
@@ -0,0 +1,981 @@
+        p2[map_orig_sorted[idx_a]] = t
+    del p
+
+    # add normalized distance matrix for current gene
+    # to full distance matrix
+    full_distance_matrix[full_distance_matrix_offset] = p2
+
+
+def cluster_distances(species_set_dict,
+                      species_set_size,
+                      hamming_distance):
+    """ Cluster gene families by species with detectable orthologs in exactly
+    the same subset of the considered species.
+
+    Ex. Assume we have 4 genes and 5 species with the following distance
+    matrix:
+
+        0             1             2             3
+    0_0 nan           nan           nan           nan
+    0_1 -1.59564844   -1.388031632  -0.9634704748 -1.342272936
+    0_2 -0.4259542606 nan           0.7035215923  1.223837777
+    0_3 -1.55041393   -1.51499567   -0.9634704748 -1.330178178
+    0_4 -0.3659762821 0.8346037464  0.6565725705  1.15274682
+    ..
+    ..
+
+    There are two binary indicator vectors to represent the species present in
+    the four genes: IIIII (gene 0, 2 and 3), II0II (gene 1). If the core set
+    threshold was 3, then there would be 1 core species set represented by
+    IIIII.
+
+    Parameters
+    ----------
+    species_set_dict: dictionary
+      dictionary containing the binary indicator vectors as
+      keys and the number of genes with identical species
+      set represented by the binary vectors as values
+    species_set_size: integer
+      threshold number of genes in a species set to
+      allow it to form a core cluster
+    hamming_distance: integer
+      maximum number of mismatches between two binary
+      indicator vectors (ex. IIII and I0II) for the
+      genes in a candidate vector to be merged into the
+      core cluster
+
+    Returns
+    -------
+    gene_clusters_dict: dictionary
+      dictionary containing core species sets as keys
+      and all belonging species sets as values
+      (determined by the Hamming distance clustering
+      algorithm)
+    """
+    sorted_species_set = sorted(species_set_dict.items(),
+                                key=operator.itemgetter(1), reverse=True)
+
+    # determine core clusters (initial species sets
+    # with more than species_set_size genes)
+    gene_clusters_dict = {}
+    # if the largest species set contains less than
+    # threshold (species_set_size) elements, set the
+    # only core cluster to the largest species set
+    if sorted_species_set[0][1] < species_set_size:
+        gene_clusters_dict[sorted_species_set[0][0]] = []
+    else:
+        for bitvector in sorted_species_set:
+            if bitvector[1] >= species_set_size:
+                gene_clusters_dict[bitvector[0]] = []
+
+    # assign species sets with fewer than species_set_size
+    # species to core clusters if the Hamming distance
+    # between the two bitvectors is less than
+    # hamming_distance
+    species_set_assigned = []
+    for cluster_core in gene_clusters_dict:
```

Comment:
In a lot of this code you could reduce the number of indexing by using temporal variables. Perhaps something like?
```
for cluster_core in gene_clusters_dict:
    bv = bitvector[0]
    if bv not in species_set_assigned and hamming(cluster_core, bv) <= hamming_distance:
         gene_clusters_dict[cluster_core].append(bv)
         species_set_assigned.append(bv)
```